File: Week_06/G20200447010024/LeetCode_208_024.js
/*
 * @lc app=leetcode.cn id=208 lang=javascript
 *
 * [208] 实现 Trie (前缀树)
 *
 * https://leetcode-cn.com/problems/implement-trie-prefix-tree/description/
 *
 * algorithms
 * Medium (63.24%)
 * Likes:    225
 * Dislikes: 0
 * Total Accepted:    28K
 * Total Submissions: 42.7K
 * Testcase Example:  '["Trie","insert","search","search","startsWith","insert","search"]\n' +
  '[[],["apple"],["apple"],["app"],["app"],["app"],["app"]]'
 *
 * 实现一个 Trie (前缀树)，包含 insert, search, 和 startsWith 这三个操作。
 * 
 * 示例:
 * 
 * Trie trie = new Trie();
 * 
 * trie.insert("apple");
 * trie.search("apple");   // 返回 true
 * trie.search("app");     // 返回 false
 * trie.startsWith("app"); // 返回 true
 * trie.insert("app");   
 * trie.search("app");     // 返回 true
 * 
 * 说明:
 * 
 * 
 * 你可以假设所有的输入都是由小写字母 a-z 构成的。
 * 保证所有输入均为非空字符串。
 * 
 * 
 */

// @lc code=start
/**
 * Initialize your data structure here.
 */
var Trie = function() {
    this.node = {}
};

/**
 * Inserts a word into the trie. 
 * @param {string} word
 * @return {void}
 */
Trie.prototype.insert = function(word) {
    const arrs = word.split('')
    let node = this.node

    for (let i = 0; i < arrs.length; i++) {
        const w = arrs[i]
        if (!node[w]) {
            node[w] = {}
        }
        node = node[w]
    }

    node.isEnd = true
};

/**
 * Returns if the word is in the trie. 
 * @param {string} word
 * @return {boolean}
 */
Trie.prototype.search = function(word) {
    const arrs = word.split('')
    let node = this.node
    
    for (let i = 0; i < arrs.length; i++) {
        if (!node[arrs[i]]) {
            return false
        }
        node = node[arrs[i]]
    }

    return !!node.isEnd
};

/**
 * Returns if there is any word in the trie that starts with the given prefix. 
 * @param {string} prefix
 * @return {boolean}
 */
Trie.prototype.startsWith = function(prefix) {
    const arrs = prefix.split('')
    let node = this.node

    for (let i = 0; i < arrs.length; i++) {
        if (!node[arrs[i]]) {
            return false
        }
        node = node[arrs[i]]
    }

    return true
};

/**
 * Your Trie object will be instantiated and called as such:
 * var obj = new Trie()
 * obj.insert(word)
 * var param_2 = obj.search(word)
 * var param_3 = obj.startsWith(prefix)
 */
// @lc code=end

File: Week_06/G20200343030024/LeetCode_208_024.java
class Trie {

    final TrieNode root = new TrieNode();

   /**
    * Initialize your data structure here.
    */
   public Trie() {
   }

   /**
    * Inserts a word into the trie.
    */
   public void insert(String word) {
       char[] wc = word.toCharArray();
       TrieNode node = root;
       for (char c : wc) {
           node = node.insert(c);
       }
       node.setEnd();
   }

   /**
    * Returns if the word is in the trie.
    */
   public boolean search(String word) {
       char[] wc = word.toCharArray();
       TrieNode node = root;
       for (char c : wc) {
           if (!node.containsKey(c)) {
               return false;
           }
           node = node.get(c);
       }

       return node.isEnd();
   }

   /**
    * Returns if there is any word in the trie that starts with the given prefix.
    */
   public boolean startsWith(String prefix) {
       char[] wc = prefix.toCharArray();
       TrieNode node = root;
       for (char c : wc) {
           if (!node.containsKey(c)) {
               return false;
           }
           node = node.get(c);
       }

       return true;
   }

   class TrieNode {
       final TrieNode[] data = new TrieNode[26];
       boolean end;

       boolean containsKey(char v) {
           return data[v - 'a'] != null;
       }

       TrieNode insert(char v) {
           TrieNode node = get(v);
           if (node == null){
               node = new TrieNode();
               data[v - 'a'] = node;
           }
           return node;
       }

       TrieNode get(char v) {
           return data[v - 'a'];
       }

       boolean isEnd() {
           return end;
       }

       void setEnd() {
           this.end = true;
       }

   }
}

/**
* Your Trie object will be instantiated and called as such:
* Trie obj = new Trie();
* obj.insert(word);
* boolean param_2 = obj.search(word);
* boolean param_3 = obj.startsWith(prefix);
*/